content/docs/11.migration-guide/plugins-discover-mechanism.md

```diff
@@ -0,0 +1,130 @@
+---
+title: Plugins Discovery Mechanism
+icon: /docs/icons/migration-guide.svg
+release: 0.17.0
+---
+
+In Kestra version 0.17.0, we decided to redesign the mechanism used to discover and load core and custom plugins.
+
+Plugins are now discovered and loaded using the standard *Java Service Loader*.
+
+## Why the Change?
+
+In earlier versions, Kestra was relying heavily on the _Bean Introspection_ mechanism provided by Micronaut Framework
+for loading plugins (as a reminder Micronaut is the main JVM-based framework used by Kestra).
+
+However, we have repeatedly encountered some limitations in maintaining backward compatibility of plugins during major
+version upgrades of Micronaut. In addition, this implementation was limiting our ability to introduce future 
+enhancements around the Kestra plugin mechanism.
+
+Thanks to this new implementation, we have been allowed not only to reduce the number of dependencies required for
+developing custom plugins, but also to divide by two the start-time of Kestra when loading a lot of plugins.
+
+Finally, this change is part of a wider effort to improve the developer experience around plugins, and
+to reduce Micronaut's exposure outside the Kestra core.
+
+Unfortunately, we've had to introduce some minor breaking-changes to the way custom plugins should be built.
+Here are the changes required to migrate to Kestra 0.17.0.
+
+## Micronaut dependencies
+
+For most plugin implementations, all Micronaut libs can be removed from the `compileOnly` dependencies in the `build.gradle` file.
+
+However, Micronaut is still required to use the utility classes provided by Kestra for running unit-tests.
+
+## Kestra's annotation processor
+
+Kestra requires a new annotation processor to be configured in the `build.gradle` file of your project (or `pom.xml` for Maven).
+
+```
+annotationProcessor group: "io.kestra", name: "processor", version: kestraVersion
+```
+
+The role of this processor is to automatically manage the `META-INF/services` file needed by Java to discover your plugins.
+
+## Custom Validations
+
+From the beginning, Kestra allows you to develop custom constraint validator using the standard Java API for bean validation (i.e: JSR-380).
+Custom validation are used to validate the properties of custom tasks.
+
+Custom validator must now implement the standard `jakarta.validation.ConstraintValidator`
+instead of the interface provided by Micronaut: `io.micronaut.validation.validator.constraints.ConstraintValidator`.
+
+In addition, custom validation annotation should now strictly adhere to the Java bean specification:
+
+Here is a migration example:
```

Comment:
```suggestion
Kestra allows you to develop a custom constraint validator using the standard Java API for bean validation (i.e., JSR-380), which is used to validate the properties of custom tasks.

::alert{type="warning"}
The custom validator must now implement the standard `jakarta.validation.ConstraintValidator`
instead of the interface provided by Micronaut: `io.micronaut.validation.validator.constraints.ConstraintValidator`.
::

In addition, custom validation annotation should now strictly adhere to the Java bean specification:

Here is a migration example:
```